tensorboard/components/tensor_widget/selection.ts

```diff
@@ -0,0 +1,340 @@
+        indices[rowDim] < rowEnd &&
+        indices[colDim] >= this.colStart &&
+        indices[colDim] < colEnd
+      ) {
+        if (status == null) {
+          status = [];
+        }
+        status.push(CellSelectionStatus.SELECTED);
+        if (indices[rowDim] === this.rowStart) {
+          status.push(CellSelectionStatus.TOP_EDGE);
+        }
+        if (indices[rowDim] === rowEnd - 1) {
+          status.push(CellSelectionStatus.BOTTOM_EDGE);
+        }
+        if (indices[colDim] === this.colStart) {
+          status.push(CellSelectionStatus.LEFT_EDGE);
+        }
+        if (indices[colDim] === colEnd - 1) {
+          status.push(CellSelectionStatus.RIGHT_EDGE);
+        }
+      }
+      return status;
+    } else {
+      throw new Error(`Unexpected length of viewDims: ${this.viewDims}`);
+    }
+  }
+
+  /**
+   * Move the selection.
+   *
+   * Updates the state of the object accordingly. It disallows going off the
+   * edges.
+   *
+   * Moving a multi-element selection always causes the selection to
+   * collapse to a single element.
+   *
+   * @param direction Direction in which this movement is being made.
+   * @return If the selection movement doesn't necessitate an update to the
+   *   slicing spec (i.e., if the new single element selection is within
+   *   the current slicing spec), `null`. Else, a new `TensorViewSlicingSpec`
+   *   object that can accommodate the new single-element selection.
+   */
+  public move(direction: SelectionMoveDirection): TensorViewSlicingSpec | null {
+    let viewRangeChanged = false;
+    if (this.rank() === 0) {
+      // No-op for a scalar.
+      return null;
+    }
+    if (
+      this.rank() === 1 &&
+      (direction === SelectionMoveDirection.LEFT ||
+        direction === SelectionMoveDirection.RIGHT)
+    ) {
+      // No-op for moving left or right in a 1D tensor.
+      return null;
+    }
+
+    if (
+      this.slicingSpec.verticalRange === null ||
+      this.slicingSpec.verticalRange[1] === null
+    ) {
+      throw new Error(`Failed to move odue to undetermined vertical range.`);
+    }
+
+    if (direction === SelectionMoveDirection.UP) {
+      if (this.rowStart > 0) {
+        this.rowStart--;
+        if (
+          this.slicingSpec.verticalRange != null &&
+          this.rowStart < this.slicingSpec.verticalRange[0]
+        ) {
+          this.slicingSpec.verticalRange[0]--;
+          this.slicingSpec.verticalRange[1]--;
+          viewRangeChanged = true;
+        }
+      }
+    } else if (direction === SelectionMoveDirection.DOWN) {
+      if (
+        this.slicingSpec.viewingDims != null &&
+        this.slicingSpec.viewingDims[0] != null &&
+        this.rowStart < this.shape[this.slicingSpec.viewingDims[0]] - 1
+      ) {
+        this.rowStart++;
+        if (
+          this.slicingSpec.verticalRange != null &&
+          this.rowStart >= this.slicingSpec.verticalRange[1]
+        ) {
+          const increment = 1;
```

Comment:
Why is this pattern (with increment pulled out as a constant) only used for DOWN?